File: python/atomr_agents/agent_host/layout.py
"""On-disk layout for an atomr-agents host root — thin facade over ``_native.host``."""

from __future__ import annotations

import os
from pathlib import Path

from atomr_agents._native import host as _host

__all__ = ["AgentPaths", "HostPaths", "default_root", "ENV_ROOT"]

ENV_ROOT = "ATOMR_HOST_ROOT"


def default_root() -> Path:
    env = os.environ.get(ENV_ROOT)
    if env:
        return Path(env).expanduser().resolve()
    return Path(_host.default_root())


class AgentPaths:
    __slots__ = ("_inner",)

    def __init__(self, root: Path | str, agent_id: str) -> None:
        if isinstance(root, _host.HostPaths):
            self._inner = root.agent(agent_id)
        elif isinstance(root, _host.AgentPaths):
            self._inner = root
        else:
            self._inner = _host.HostPaths(root=str(root)).agent(agent_id)

    @classmethod
    def _wrap(cls, inner: object) -> AgentPaths:
        obj = cls.__new__(cls)
        obj._inner = inner
        return obj

    @property
    def root(self) -> Path:
        return Path(self._inner.root)

    @property
    def agent_id(self) -> str:
        return self._inner.agent_id

    @property
    def dir(self) -> Path:
        return Path(self._inner.dir)

    @property
    def agent_yaml(self) -> Path:
        return Path(self._inner.agent_yaml)

    @property
    def soul_md(self) -> Path:
        return Path(self._inner.soul_md)

    @property
    def rules_md(self) -> Path:
        return Path(self._inner.rules_md)

    @property
    def memory_md(self) -> Path:
        return Path(self._inner.memory_md)

    @property
    def user_md(self) -> Path:
        return Path(self._inner.user_md)

    @property
    def skills_dir(self) -> Path:
        return Path(self._inner.skills_dir)

    @property
    def hooks_dir(self) -> Path:
        return Path(self._inner.hooks_dir)

    @property
    def state_dir(self) -> Path:
        return Path(self._inner.state_dir)

    @property
    def threads_dir(self) -> Path:
        return Path(self._inner.threads_dir)

    @property
    def checkpoints_dir(self) -> Path:
        return Path(self._inner.checkpoints_dir)

    @property
    def memory_db(self) -> Path:
        return Path(self._inner.memory_db)

    def ensure(self) -> None:
        self._inner.ensure()

    def __eq__(self, other: object) -> bool:
        if isinstance(other, AgentPaths):
            return self.root == other.root and self.agent_id == other.agent_id
        return NotImplemented

    def __hash__(self) -> int:
        return hash((self.root, self.agent_id))

    def __repr__(self) -> str:
        return f"AgentPaths(root={self.root!r}, agent_id={self.agent_id!r})"


class HostPaths:
    __slots__ = ("_inner",)

    def __init__(self, root: Path | str) -> None:
        self._inner = _host.HostPaths(root=str(root))

    @classmethod
    def _wrap(cls, inner: object) -> HostPaths:
        obj = cls.__new__(cls)
        obj._inner = inner
        return obj

    @property
    def root(self) -> Path:
        return Path(self._inner.root)

    @property
    def config_yaml(self) -> Path:
        return Path(self._inner.config_yaml)

    @property
    def agents_md(self) -> Path:
        return Path(self._inner.agents_md)

    @property
    def agents_dir(self) -> Path:
        return Path(self._inner.agents_dir)

    @property
    def channels_dir(self) -> Path:
        return Path(self._inner.channels_dir)

    @property
    def crons_dir(self) -> Path:
        return Path(self._inner.crons_dir)

    @property
    def tools_dir(self) -> Path:
        return Path(self._inner.tools_dir)

    @property
    def registry_dir(self) -> Path:
        return Path(self._inner.registry_dir)

    @property
    def events_jsonl(self) -> Path:
        return Path(self._inner.events_jsonl)

    def agent(self, agent_id: str) -> AgentPaths:
        return AgentPaths._wrap(self._inner.agent(agent_id))

    def list_agent_ids(self) -> list[str]:
        return list(self._inner.list_agent_ids())

    def ensure(self) -> None:
        self._inner.ensure()

    def __eq__(self, other: object) -> bool:
        if isinstance(other, HostPaths):
            return self.root == other.root
        return NotImplemented

    def __hash__(self) -> int:
        return hash(self.root)

    def __repr__(self) -> str:
        return f"HostPaths(root={self.root!r})"

File: crates/host/src/layout.rs
//! On-disk layout for an atomr-agents host root.
//!
//! Default root: `~/.atomr/host/` (overridable via `$ATOMR_HOST_ROOT`).

use std::env;
use std::path::{Path, PathBuf};

use crate::error::{HostError, HostResult};

pub const ENV_ROOT: &str = "ATOMR_HOST_ROOT";

/// Resolve the default host root, honoring `$ATOMR_HOST_ROOT`.
///
/// Does **not** create the directory; callers (`init` and
/// [`HostPaths::ensure`]) own creation so cold lookups stay read-only.
pub fn default_root() -> PathBuf {
    if let Ok(env) = env::var(ENV_ROOT) {
        let p = PathBuf::from(env);
        return expand_user(&p);
    }
    let home = dirs_home();
    home.join(".atomr").join("host")
}

fn dirs_home() -> PathBuf {
    if let Ok(home) = env::var("HOME") {
        return PathBuf::from(home);
    }
    // Final fallback — useful in CI / sandboxes that strip HOME.
    PathBuf::from("/tmp")
}

fn expand_user(p: &Path) -> PathBuf {
    let s = p.to_string_lossy();
    if let Some(rest) = s.strip_prefix("~/") {
        return dirs_home().join(rest);
    }
    if s == "~" {
        return dirs_home();
    }
    p.to_path_buf()
}

/// Filesystem layout for the host root.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct HostPaths {
    pub root: PathBuf,
}

impl HostPaths {
    pub fn new(root: impl Into<PathBuf>) -> Self {
        let root = expand_user(&root.into());
        Self { root }
    }

    pub fn config_yaml(&self) -> PathBuf {
        self.root.join("config.yaml")
    }

    pub fn agents_md(&self) -> PathBuf {
        self.root.join("AGENTS.md")
    }

    pub fn agents_dir(&self) -> PathBuf {
        self.root.join("agents")
    }

    pub fn channels_dir(&self) -> PathBuf {
        self.root.join("channels")
    }

    pub fn crons_dir(&self) -> PathBuf {
        self.root.join("crons")
    }

    pub fn tools_dir(&self) -> PathBuf {
        self.root.join("tools")
    }

    pub fn registry_dir(&self) -> PathBuf {
        self.root.join("registry")
    }

    pub fn evals_dir(&self) -> PathBuf {
        self.root.join("evals")
    }

    pub fn events_jsonl(&self) -> PathBuf {
        self.root.join("events.jsonl")
    }

    pub fn mcp_dir(&self) -> PathBuf {
        self.root.join("mcp")
    }

    pub fn agent(&self, agent_id: &str) -> AgentPaths {
        AgentPaths {
            root: self.root.clone(),
            agent_id: agent_id.to_string(),
        }
    }

    /// Enumerate agent ids by scanning `agents/` for sub-directories.
    /// Returns an empty list when the host root does not exist yet.
    pub fn list_agent_ids(&self) -> Vec<String> {
        let agents = self.agents_dir();
        if !agents.is_dir() {
            return Vec::new();
        }
        let mut out: Vec<String> = match std::fs::read_dir(&agents) {
            Ok(rd) => rd
                .filter_map(|e| e.ok())
                .filter(|e| e.file_type().map(|t| t.is_dir()).unwrap_or(false))
                .filter_map(|e| e.file_name().into_string().ok())
                .filter(|name| !name.starts_with('.'))
                .collect(),
            Err(_) => Vec::new(),
        };
        out.sort();
        out
    }

    /// Create the host directory skeleton (idempotent).
    pub fn ensure(&self) -> HostResult<()> {
        for d in [
            self.root.clone(),
            self.agents_dir(),
            self.channels_dir(),
            self.crons_dir(),
            self.tools_dir(),
            self.registry_dir(),
        ] {
            std::fs::create_dir_all(&d).map_err(|e| HostError::io(&d, e))?;
        }
        Ok(())
    }
}

/// Filesystem layout for a single agent under `<root>/agents/<id>/`.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct AgentPaths {
    pub root: PathBuf,
    pub agent_id: String,
}

impl AgentPaths {
    pub fn dir(&self) -> PathBuf {
        self.root.join("agents").join(&self.agent_id)
    }
    pub fn agent_yaml(&self) -> PathBuf {
        self.dir().join("agent.yaml")
    }
    pub fn soul_md(&self) -> PathBuf {
        self.dir().join("SOUL.md")
    }
    pub fn rules_md(&self) -> PathBuf {
        self.dir().join("RULES.md")
    }
    pub fn memory_md(&self) -> PathBuf {
        self.dir().join("MEMORY.md")
    }
    pub fn user_md(&self) -> PathBuf {
        self.dir().join("USER.md")
    }
    pub fn skills_dir(&self) -> PathBuf {
        self.dir().join("skills")
    }
    pub fn hooks_dir(&self) -> PathBuf {
        self.dir().join("hooks")
    }
    pub fn state_dir(&self) -> PathBuf {
        self.dir().join("state")
    }
    pub fn threads_dir(&self) -> PathBuf {
        self.state_dir().join("threads")
    }
    pub fn checkpoints_dir(&self) -> PathBuf {
        self.state_dir().join("checkpoints")
    }
    pub fn memory_db(&self) -> PathBuf {
        self.state_dir().join("memory.db")
    }

    pub fn ensure(&self) -> HostResult<()> {
        for d in [
            self.dir(),
            self.skills_dir(),
            self.hooks_dir(),
            self.state_dir(),
            self.threads_dir(),
            self.checkpoints_dir(),
        ] {
            std::fs::create_dir_all(&d).map_err(|e| HostError::io(&d, e))?;
        }
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::tempdir;

    #[test]
    fn host_paths_basics() {
        let tmp = tempdir().unwrap();
        let paths = HostPaths::new(tmp.path());
        assert_eq!(paths.config_yaml(), tmp.path().join("config.yaml"));
        assert_eq!(paths.agents_dir(), tmp.path().join("agents"));
        assert_eq!(paths.events_jsonl(), tmp.path().join("events.jsonl"));
    }

    #[test]
    fn agent_paths_under_root() {
        let tmp = tempdir().unwrap();
        let paths = HostPaths::new(tmp.path()).agent("alpha");
        assert_eq!(paths.agent_yaml(), tmp.path().join("agents/alpha/agent.yaml"));
        assert_eq!(paths.soul_md(), tmp.path().join("agents/alpha/SOUL.md"));
        assert_eq!(
            paths.checkpoints_dir(),
            tmp.path().join("agents/alpha/state/checkpoints")
        );
    }

    #[test]
    fn ensure_idempotent() {
        let tmp = tempdir().unwrap();
        let paths = HostPaths::new(tmp.path());
        paths.ensure().unwrap();
        paths.ensure().unwrap();
        assert!(paths.agents_dir().is_dir());
    }

    #[test]
    fn list_agent_ids_filters_hidden() {
        let tmp = tempdir().unwrap();
        let paths = HostPaths::new(tmp.path());
        paths.ensure().unwrap();
        std::fs::create_dir_all(paths.agents_dir().join("alpha")).unwrap();
        std::fs::create_dir_all(paths.agents_dir().join(".hidden")).unwrap();
        std::fs::create_dir_all(paths.agents_dir().join("beta")).unwrap();
        assert_eq!(paths.list_agent_ids(), vec!["alpha", "beta"]);
    }
}
